pkg/nsx/services/securitypolicy/store.go

```diff
@@ -1,153 +1,132 @@
-			if err != nil {
-				for _, e := range err {
-					fatalErrors <- e
-				}
-			}
-			c, _ := a.(model.Group)
-			err2 := service.GroupStore.Add(c)
-			if err2 != nil {
-				fatalErrors <- err2
-			}
-		}
-		cursor = response.Cursor
-		if cursor == nil {
-			break
-		}
-		c, _ := strconv.Atoi(*cursor)
-		if int64(c) >= *response.ResultCount {
-			break
+var key = util.TagScopeSecurityPolicyCRUID
+
+func keyFuncSecurityPolicy(obj interface{}) (string, error) {
+	o := obj.(model.SecurityPolicy)
+	return *o.Id, nil
+}
+
+func keyFuncGroup(obj interface{}) (string, error) {
+	o := obj.(model.Group)
+	return *o.Id, nil
+}
+
+func keyFuncRule(obj interface{}) (string, error) {
+	o := obj.(model.Rule)
+	return *o.Id, nil
+}
+
+func indexFuncSecurityPolicy(obj interface{}) ([]string, error) {
+	o := obj.(model.SecurityPolicy)
+	return filterTag(o.Tags), nil
+}
+
+func indexFuncGroup(obj interface{}) ([]string, error) {
+	o := obj.(model.Group)
+	return filterTag(o.Tags), nil
+}
+
+func indexFuncRule(obj interface{}) ([]string, error) {
+	o := obj.(model.Rule)
+	return filterTag(o.Tags), nil
+}
+
+var filterTag = func(v []model.Tag) []string {
+	res := make([]string, 0, 5)
+	for _, tag := range v {
+		if *tag.Scope == key {
+			res = append(res, *tag.Tag)
 		}
 	}
+	return res
 }
 
-func querySecurityPolicy(service *SecurityPolicyService, wg *sync.WaitGroup, fatalErrors chan error) {
-	defer wg.Done()
-	queryParam := fmt.Sprintf("%s:%s", common.ResourceType, common.ResourceTypeSecurityPolicy) + " AND " + queryTagCondition(service)
-	var cursor *string = nil
-	pageSize := common.PageSize
-	for {
-		response, err := service.NSXClient.QueryClient.List(queryParam, cursor, nil, &pageSize, nil, nil)
-		err = common.TransError(err)
-		if _, ok := err.(nsxutil.PageMaxError); ok == true {
-			common.DecrementPageSize(&pageSize)
-			continue
-		}
-		if err != nil {
-			fatalErrors <- err
-		}
-		for _, g := range response.Results {
-			a, err := Converter.ConvertToGolang(g, model.SecurityPolicyBindingType())
-			if err != nil {
-				for _, e := range err {
-					fatalErrors <- e
-				}
-			}
-			c, _ := a.(model.SecurityPolicy)
-			err2 := service.SecurityPolicyStore.Add(c)
-			if err2 != nil {
-				fatalErrors <- err2
-			}
-		}
-		cursor = response.Cursor
-		if cursor == nil {
-			break
-		}
-		c, _ := strconv.Atoi(*cursor)
-		if int64(c) >= *response.ResultCount {
-			break
+func InitializeStore(service *SecurityPolicyService) {
+	service.Store = make(map[string]cache.Indexer)
+	service.Store[ResourceTypeSecurityPolicy] = cache.NewIndexer(keyFuncSecurityPolicy, cache.Indexers{key: indexFuncSecurityPolicy})
+	service.Store[ResourceTypeGroup] = cache.NewIndexer(keyFuncGroup, cache.Indexers{key: indexFuncGroup})
+	service.Store[ResourceTypeRule] = cache.NewIndexer(keyFuncRule, cache.Indexers{key: indexFuncRule})
+}
+
+type OperateSecurityPolicyStore struct {
+	securityService *SecurityPolicyService
```

Comment:
name it SecurityPolicyStore? and store should not refer to a service, it should just refer to Indexer (embedded a Indexer)
RuleStore, GroupStore the same